provider/common/bootstrap_test.go

```diff
@@ -398,3 +398,53 @@ func (s *BootstrapSuite) TestWaitSSHRefreshAddresses(c *gc.C) {
 		"Waiting for address\n"+
 			"(.|\n)*(Attempting to connect to 0.1.2.4:22\n)+(.|\n)*")
 }
+
+type FormatHardwareSuite struct{}
+
+var _ = gc.Suite(&FormatHardwareSuite{})
+
+func (s *FormatHardwareSuite) check(c *gc.C, hw *instance.HardwareCharacteristics, expected string) {
+	c.Check(common.FormatHardware(hw), gc.Equals, expected)
+}
+
+func (s *FormatHardwareSuite) TestNil(c *gc.C) {
+	s.check(c, nil, "")
+	s.check(c, &instance.HardwareCharacteristics{}, "")
```

Comment:
Actually it's not the same. TestNil checks what happens when the *HardwareCharacteristics is nil and when there is an instance but all the field values are nil.

The other tests below checks what happens when the fields have a value but they point to the value's zero value.

To make things clearer, I might split up TestNil so that there's TestNil and TestFieldsNil.